core/client/grpcc/grpc_client.go

```diff
@@ -225,70 +223,162 @@ func (c *Client) ListHistory(ctx context.Context, userID, appID string, start, e
 	return profiles, nil
 }
 
-// Update creates an UpdateEntryRequest for a user, attempt to submit it multiple
-// times depending on RetryCount.
-func (c *Client) Update(ctx context.Context, appID, userID string, profileData []byte,
-	signers []signatures.Signer, authorizedKeys []*keyspb.PublicKey,
-	opts ...grpc.CallOption) (*entry.Mutation, error) {
-	getResp, err := c.cli.GetEntry(ctx, &pb.GetEntryRequest{
-		DomainId:      c.domainID,
-		UserId:        userID,
-		AppId:         appID,
-		FirstTreeSize: c.trusted.TreeSize,
-	}, opts...)
-	if err != nil {
-		return nil, fmt.Errorf("GetEntry(%v): %v", userID, err)
+// Update creates an UpdateEntryRequest for a user,
+// attempt to submit it multiple times depending on RetryCount.
+func (c *Client) Update(ctx context.Context, u *tpb.User, signers []signatures.Signer) (*entry.Mutation, error) {
+	if got, want := u.DomainId, c.domainID; got != want {
+		return nil, fmt.Errorf("u.DomainID: %v, want %v", got, want)
 	}
-	Vlog.Printf("Got current entry...")
-
-	if err := c.kt.VerifyGetEntryResponse(ctx, c.domainID, appID, userID, &c.trusted, getResp); err != nil {
-		return nil, fmt.Errorf("VerifyGetEntryResponse(): %v", err)
+	// 1. pb.User + ExistingEntry -> Mutation
+	m, err := c.newMutation(ctx, u)
+	if err != nil {
+		return nil, err
 	}
 
-	m, err := c.kt.NewMutation(c.domainID, appID, userID, profileData, authorizedKeys,
-		getResp.GetVrfProof(), getResp.GetLeafProof().GetLeaf().GetLeafValue())
-	if err != nil {
-		return nil, fmt.Errorf("CreateUpdateEntryRequest: %v", err)
+	if err := c.QueueMutation(ctx, m, signers); err != nil {
+		return nil, err
 	}
 
-	err = c.Retry(ctx, m, signers, opts...)
-	// Retry submitting until an inclusion proof is returned.
-	for i := 0; err == ErrRetry && i < c.RetryCount; i++ {
-		time.Sleep(c.RetryDelay)
-		err = c.Retry(ctx, m, signers, opts...)
+	// 3. Wait for update
+	m, err = c.WaitForUserUpdate(ctx, m)
+	for i := 0; i < c.RetryCount; i++ {
```

Comment:
Might be better to retry using the deadline on context than a retry count.

Reply:
Good idea.  I'll remove RetryCount all together

Reply:
ok, that took a bit longer than expected, but it's done now.